### DISCUSSION TITLE
Change historacal charge price

### DISCUSSION THREAD

**Discussion by Bolten88** [teslamate-org/teslamate#3387] [General]
Hello,

i accidentally set the charging price of a geofence location to €30 instead of €0.30. The total price in grafana of charging at that location are completely wrong now. Can i change the price of the previous charging sessions from that location?

Kind regards,
Mark

**Comment by DrMichael** [comment]
I am sure, that that can be done with some database magic...

**Comment by sparvata** [comment]
get the id from the charges table and you can do like the below if the charge id is 9 that you want to modify the cost for. In the query below, if you set the cost to NULL instead of a value, you can edit it from the web UI I think.

```
docker-compose exec database psql teslamate teslamate << .
SELECT * FROM charging_processes where id = 9;
update charging_processes set cost = 0.30 where id = '9';
.
```

**Comment by DrMichael** [comment]
Ok, that challenged me...
Assuming your geofence is named 'Zuhause', you can divide all charging prices at that location by 100 with:
```
update charging_processes
	set cost = cost/100
from geofences
where charging_processes.geofence_id = geofences.id AND geofences.name='Zuhause'
```
(use that instead of line 2 and 3 above)

**Comment by sparvata** [comment]
note that, if you want to make the change for all charges at that geolocation you use the geofence_id. If you need to change for a specific charge, you can use the id from the charges table.

**Comment by DrMichael** [comment]
And don't forget to make a backup before...

**Comment by Bolten88** [comment]
Thanks All, this is going to work. Thanks for pointing me in the right direction. And sure i do have backups :)

**Comment by damccull** [comment]
I just started looking into this. It looks like price per kWh is calculated realtime, while the database stores the total cost and energy added. Is that correct? Going to need that info to make the correct updates, because it's looking like I'll need a query like this:
```sql
UPDATE charging_processes SET cost = charge_energy_added * 0.39 WHERE geofence_id=1 AND start_date >= '2023-08-01 00:00:00.000' AND start_date <= '2023-08-31 23:59:59.999';
```
Is this correct, assuming I want to update all August 2023 charges to 39 cents per kWh?

**Reply by khorton** [reply]
This worked perfectly for me.  But, the cost calculation should be based on the electricity used, which is a bit more than the electricity added.

`
docker compose exec -T database psql -d teslamate -U teslamate << . 
UPDATE charging_processes SET cost = charge_energy_used * 0.39 WHERE geofence_id=1 AND start_date >= '2023-08-01 00:00:00.000' AND start_date <= '2023-08-31 23:59:59.999';
`

**Comment by giovaorama** [comment]
Hi all,
this topic is really important for me since I really know the actual cost pro kWh of my home charging session only a couple of month after, when I get the bill from my energy supplier.

I made some tests and I see a small glitch that may lead to zero costs for some sessions. 

For some "very short session" I have at DB records with a valid _charge_energy_added_ value but **charge_energy_used is null** (I do not know why, but in graphana these are showed correctly)

To solve this I think is is better to use this formula in the UPDATE clause

`cost=GREATEST(charge_energy_added,charge_energy_used)*0.39`.

Moreover we can simplify the start_date selection using like

For instance to update all the charges at my home geofence made in May 2025 where 

`
UPDATE charging_processes SET cost=GREATEST(charge_energy_used,charge_energy_added)*0.229 WHERE geofence_id='1' AND start_date::text LIKE '%2025-05%';
`

I test it in my installation and all works correctly

**Comment by JakobLichterfeld** [comment]
> Hello,
> 
> i accidentally set the charging price of a geofence location to €30 instead of €0.30. The total price in grafana of charging at that location are completely wrong now. Can i change the price of the previous charging sessions from that location?
> 
> Kind regards, Mark

For sure, just change it back in geofence and it will ask you if you will change it in past as well. If not directly, change to 0 and then to new value.

**Reply by fran-6co** [reply]
Hi @JakobLichterfeld ,

I came to search for this kind of scenario. Recently, I changed my electrical bill, and now it costs less per kWh; however, it is not clear the process for people in the same situation. Let me walk you through the steps I followed:

1. You want to update the per kWh rate on the geofence.
2. I went to the geofence menu to see if there is the possibility.
3. The menu looked like if you change that, it will change for all the charges in the geofence. So I came here to look for a solution or write an issue.

Possible solution I was expecting to see: click on a button that says "new pricing?" and that asks you:

- Update ALL previous charges in this geofence.
- Start a new pricing from a date.

Let me know if you think this might be better as an issue.

Thank you.

**Reply by JakobLichterfeld** [reply]
> just change it (...) in geofence and it will ask you if you will change it in past

**Reply by fran-6co** [reply]
I just changed it and there was no message for me...
